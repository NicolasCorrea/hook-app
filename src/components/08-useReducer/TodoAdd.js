import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddTodo }) => {
  
  const [{ description }, setFormState, reset] = useForm({
    description: ""
  });
  
  
  const handleSubmit = (e) => {
    e.preventDefault()
    if (description.length < 1) {
      return;
    }
    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };
    handleAddTodo(newToDo)
    reset();
  }
  
  
  return (
    <>
      <h4>Add ToDo</h4>
      <hr />

      <form onSubmit={handleSubmit}>
        <div className="form-group">
          <input type="text"
            className="form-control"
            name="description"
            id="description"
            placeholder="Add Task"
            autoComplete="off"
            value={description}
            onChange={setFormState}
          />
        </div>
        <button
          type="submit"
          className="btn btn-block btn-primary"
        >
          Add
            </button>

      </form>
    </>
  )
}
