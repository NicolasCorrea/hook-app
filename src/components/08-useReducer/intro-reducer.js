
const initialState = [
  {
    id: 1,
    todo: "Buy Bread",
    done: false,
  }
];

const todoReducer = (state = initialState, action) => {
  
  switch (action?.type) {
    case "agregar":
      return [...state, action.payload];
      break;
  
    default:
      break;
  }
    
  
  return state
}

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "Buy milk",
  done: false,
}

const action = {
  type: "agregar",
  payload: newTodo
}

todos = todoReducer(todos, action)



console.log(todos)