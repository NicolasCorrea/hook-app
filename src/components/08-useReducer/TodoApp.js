import { useEffect, useReducer } from "react"
import { TodoAdd } from "./TodoAdd"
import { TodoList } from "./TodoList"
import { todoReducer } from "./todoReducer"
import "./style.css"




const init = () => {
  return JSON.parse(localStorage.getItem("todos")) || [];
}

export const TodoApp = () => {

  const [todos, dispatch] = useReducer(todoReducer, [], init)

  const handleAddTodo = (newToDo) => {
    dispatch({
      type: "add",
      payload: newToDo
    });
  }

  const handleDelete = todoId => {
    const action = {
      type: "delete",
      payload: todoId
    }
    dispatch(action);
  }

  const handleToggle = todoId => {
    dispatch({
      type: "toggle",
      payload: todoId,
    })
  }

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos])

  return (
    <div>
      <h1>ToDo App (<small>{todos.length}</small>)</h1>
      <hr />

      <div className="row">

        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>

      </div>

    </div>
  )
}
