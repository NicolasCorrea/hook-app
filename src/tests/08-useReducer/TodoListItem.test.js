import { shallow } from "enzyme"
import { TodoListItem } from "../../components/08-useReducer/TodoListItem"
import { demoTodos } from "../fixtures/demoTodos"

describe('<TodoListItem />', () => {
  const handleDelete= jest.fn();
  const handleToggle= jest.fn();
  const index = 1;
  const wrapper = shallow(<TodoListItem
    todo={demoTodos[0]}
    handleDelete={handleDelete}
    handleToggle={handleToggle}
    index={index}
  />)

  test('should render', () => {
    expect(wrapper).toMatchSnapshot();
  })
  test('should call handleDelete function', () => {
    wrapper.find('button').simulate('click');
    expect(handleDelete).toHaveBeenCalledWith(demoTodos[0].id)
  })
  test('should call handleToggle', () => {
    wrapper.find('p').simulate('click');
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id)
  })

  test('should show the text succefully', () => {
    const p = wrapper.find('p').text().trim();
    expect(p).toBe(`${index+1}. ${demoTodos[0].desc}`);
  })


  test('should have complete class when a todo is complete', () => {
    const wrapper = shallow(<TodoListItem
      todo={{...demoTodos[0], done: true}}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
      index={index}
    />)

    const p = wrapper.find('p');
    expect(p.hasClass('complete')).toBe(true);
  
  })
  
})
