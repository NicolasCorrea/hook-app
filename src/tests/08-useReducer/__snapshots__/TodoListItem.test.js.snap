// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListItem /> should render 1`] = `
<li
  className="list-group-item "
>
  <p
    className=""
    onClick={[Function]}
  >
    2
    . 
    Aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
